README.md
# animated_drag_gridview

该插件的灵感来源于DragableGridview 和 reorderable_grid_view， DragableGridview 已经很久没有更新了也不支持空安全版本，DragableGridview 使用 transform 来控制拖拽，拖拽的是本身自己，由于Gridview 渲染层级的原因，前面的item拖向后面的item 会被后面的item覆盖掉。而reorderable_grid_view 解决了这个层级的问题，但作者使用的是OverlayEntry，这将导致拖拽的小部件悬浮于一切之上，包括tabbar,使用时感觉很不舒服。
所以 animated_drag_gridview 使用的Stack + Positioned，这样可以避免这些问题。

[DragableGridview](https://github.com/baoolong/DragableGridview)

[reorderable_grid_view](https://github.com/huhuang03/reorderable_grid_view)


## Usage

Add this to your package's pubspec.yaml file:

	dependencies:
	  
    animated_drag_gridview:
      git:
        url: https://github.com/zhou-Flutter/animated_drag_gridview.git
        ref: master

## Import it

	import 'package:animated_drag_gridview/animated_drag_gridview.dart';

## Example
<div>
    <img width="28%" height="28%" src="https://raw.githubusercontent.com/zhou-Flutter/animated_drag_gridview/master/example/assets/example_01.jpg"/>
    <img width="28%" height="28%" src="https://raw.githubusercontent.com/zhou-Flutter/animated_drag_gridview/master/example/assets/example_02.gif"/>
</div>


    import 'package:flutter/material.dart';
    import 'package:animated_drag_gridview/animated_drag_gridview.dart';

    class DragGridView extends StatefulWidget {
    DragGridView({Key? key}) : super(key: key);

    @override
    State<DragGridView> createState() => _DragGridViewState();
    }

    class _DragGridViewState extends State<DragGridView> {
    bool isActive = false;
    @override
    void initState() {
        super.initState();
    }

    List items = [
        "张飞",
        '关羽',
        '李白',
        '刘备',
        '周瑜',
        '扁鹊',
        '上官',
        '苏烈',
        '元哥',
        '曹操',
        '廉颇',
        '夏侯',
        '梦琪',
    ];
    @override
    Widget build(BuildContext context) {
        return Scaffold(
        backgroundColor: Colors.white,
        appBar: AppBar(
            centerTitle: true,
            backgroundColor: Colors.white,
            elevation: 0.5,
            title: Text("频道设置", style: TextStyle(color: Colors.black)),
        ),
        body: Column(
            children: [
            topEdit(),
            AnimateDragGridView(
                crossAxisCount: 4,
                childAspectRatio: 2.2,
                crossAxisSpacing: 8.0,
                mainAxisSpacing: 8.0,
                isOpenDragAble: true,
                isActive: isActive,
                fixedNum: 3,
                animationDuration: 350, //milliseconds
                longPressDuration: 600, //milliseconds
                padding: const EdgeInsets.symmetric(horizontal: 15),
                physics: const NeverScrollableScrollPhysics(),
                children: items.map((e) => buildItem(e)).toList(),
                tRWidget: tRWidget(),
                onReorder: (oldIndex, newIndex) {
                var removeItem = items.removeAt(oldIndex);
                items.insert(newIndex, removeItem);
                setState(() {});
                },
                onDeleteItem: (index) {
                items.removeAt(index);
                setState(() {});
                },
                onActive: (isActive) {
                this.isActive = isActive;
                setState(() {});
                },
                onTapItem: (index) {
                print("点击的item");
                print(index);
                },
            ),
            ],
        ),
        );
    }

    Widget tRWidget() {
        return Container(
        height: 15,
        width: 15,
        alignment: Alignment.center,
        decoration: BoxDecoration(
            color: Color.fromARGB(255, 208, 208, 220),
            borderRadius: BorderRadius.circular(50),
        ),
        child: Icon(
            Icons.remove,
            color: Colors.white,
            size: 13,
        ),
        );
    }

    Widget buildItem(e) {
        return Container(
        decoration: BoxDecoration(
            color: Color.fromARGB(255, 231, 236, 245),
            borderRadius: BorderRadius.circular(2),
        ),
        alignment: Alignment.center,
        child: Text(
            e.toString(),
            style: TextStyle(
            color: Color.fromARGB(255, 86, 86, 90),
            ),
        ),
        );
    }

    Widget topEdit() {
        return Container(
        padding: EdgeInsets.all(15),
        child: Row(
            crossAxisAlignment: CrossAxisAlignment.end,
            children: [
            Text(
                "我的频道",
                style: TextStyle(
                    fontSize: 17, fontWeight: FontWeight.w600, color: Colors.black),
            ),
            SizedBox(width: 5),
            Text(
                isActive ? "长按可拖拽排序" : "点击进入频道",
                style: TextStyle(fontSize: 15, color: Colors.black26),
            ),
            Spacer(),
            InkWell(
                child: Text(
                isActive ? "完成" : "编辑",
                style: TextStyle(fontSize: 15, color: Colors.blue),
                ),
                onTap: () {
                isActive = !isActive;
                setState(() {});
                },
            ),
            ],
        ),
        );
    }
    }



CHANGELOG.md
## 0.0.1

* TODO: 可拖拽排序的网格布局.


lib/drag_info.dart
import 'package:animated_drag_gridview/animated_drag_gridview.dart';
import 'package:animated_drag_gridview/drag_gridview_mixin.dart';
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';

class DragInfo extends StatefulWidget {
  int animationDuration;
  DragInfo({
    required this.animationDuration,
    Key? key,
  }) : super(key: key);

  @override
  State<DragInfo> createState() => DragInfoState();
}

class DragInfoState extends State<DragInfo> with TickerProviderStateMixin {
  Offset updateOffset = Offset(0, 0); //移动的位置
  Offset dragInfoHome = Offset(0, 0); //目标位置

  bool isOnPaning = false;
  late DragGridViewMixin _listState;

  AnimationController? _dragInfoAnimation; //动画控制

  AnimationController? _controller; //size 大小动画

  Widget itemChild = const SizedBox();
  Widget tRWidget = const SizedBox(); //右上角 小组件
  double itemWidth = 0;
  double itemHeight = 0;

  @override
  void initState() {
    super.initState();
    _listState = DragGridViewMixin.of(context);
    _listState.dragInfo(this);
  }

  dragInfoStart(
    Offset updateOffset,
    Widget itemChild,
    Widget tRWidget,
    double itemWidth,
    double itemHeight,
  ) {
    this.itemChild = itemChild;
    this.updateOffset = updateOffset;
    this.itemWidth = itemWidth;
    this.itemHeight = itemHeight;
    this.tRWidget = tRWidget;
    isOnPaning = true;
    _controller = AnimationController(
        duration: const Duration(milliseconds: 100), vsync: this)
      ..forward();

    setState(() {});
  }

  dragInfoUpdate(Offset offset) {
    updateOffset = offset;
    setState(() {});
  }

  dragInfoCancel(currentIndex, toIndex, dragInfoOffset) {
    _controller!.reset();
    _controller!.dispose();
    //计算 dragInfo 落脚点
    dragInfoHome = dragInfoOffset;

    //动画控制器

    _dragInfoAnimation = AnimationController(vsync: this)
      ..duration = Duration(milliseconds: widget.animationDuration)
      ..addListener(rebuild)
      ..addStatusListener((status) {
        if (status == AnimationStatus.completed) {
          _dragInfoAnimation?.dispose();
          _dragInfoAnimation = null;
          //重置当前状态
          updateOffset = Offset.zero;
          isOnPaning = false;
          _listState.dragCancalBack();
          setState(() {});
        }
      })
      ..forward(from: 0.0);
  }

  Offset get dragInfoOffset {
    if (_dragInfoAnimation != null) {
      return Offset.lerp(
        updateOffset,
        dragInfoHome,
        Curves.easeInOut.transform(_dragInfoAnimation!.value),
      )!;
    }
    return updateOffset;
  }

  @override
  Widget build(BuildContext context) {
    return isOnPaning == false
        ? const SizedBox()
        : Positioned(
            top: updateOffset.dy,
            left: updateOffset.dx,
            child: ScaleTransition(
              scale: Tween(begin: 1.0, end: 1.1).animate(_controller!),
              child: Stack(
                clipBehavior: Clip.none,
                children: [
                  SizedBox(
                    width: itemWidth,
                    height: itemHeight,
                    child: itemChild,
                  ),
                  Positioned(
                    top: -2,
                    right: -2,
                    child: tRWidget,
                  )
                ],
              ),
            ),
          );
  }

  //重新 build
  void rebuild() {
    if (mounted) {
      updateOffset = dragInfoOffset;
      setState(() {});
    }
  }
}


lib/drag_item.dart
import 'package:animated_drag_gridview/animated_drag_gridview.dart';
import 'package:animated_drag_gridview/drag_gridview_mixin.dart';
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:flutter/src/foundation/key.dart';
import 'package:flutter/src/widgets/framework.dart';
import 'package:vibration/vibration.dart';

class DragItem extends StatefulWidget {
  int index;
  int fixedNum;
  int crossAxisCount;
  double mainAxisSpacing;
  double crossAxisSpacing;

  bool isOpenDragAble;
  bool isActive;
  int animationDuration;
  int longPressDuration;

  Widget tRWidget;
  Widget itemChild;

  GestureCallback gestureCallback;

  DragItem({
    required this.index,
    required this.gestureCallback,
    required this.itemChild,
    required this.mainAxisSpacing,
    required this.crossAxisSpacing,
    required this.crossAxisCount,
    required this.isOpenDragAble,
    required this.isActive,
    required this.fixedNum,
    required this.tRWidget,
    required this.animationDuration,
    required this.longPressDuration,
    Key? key,
  }) : super(key: key);

  @override
  State<DragItem> createState() => DragItemState();
}

class DragItemState extends State<DragItem> with TickerProviderStateMixin {
  bool _dragging = false; //是否在拖拽

  bool isLongPress = false; //长按控制

  int get index => widget.index;

  late DragGridViewMixin _listState;

  AnimationController? _offsetAnimation;

  late int currentIndex; //当前 index

  int newIndex = -1;

  Offset _startOffset = Offset.zero; //开始的偏移位置

  Offset _targetOffset = Offset.zero; //目标偏移位置

  bool isAnimateEnd = false; //用来判断拖拽结束时候 是否有动画

  bool isAnimate = false; //用来限制删除动画时，不能点击其他item

  @override
  void initState() {
    super.initState();
    currentIndex = widget.index;
    _listState = DragGridViewMixin.of(context);
    _listState.dragItem(this);
  }

  @override
  void didChangeDependencies() {
    super.didChangeDependencies();
    _listState.dragItem(this);
  }

  //拖拽更新 int oldIndex, int newIndex
  void updateForGap(dragIndex, newIndex) {
    this.newIndex = newIndex;

    if (!mounted) return;

    if (_dragging) {
      return;
    }
    if (index < widget.fixedNum) return;

    int indexOffsetLeg = dragIndex - newIndex;

    for (int i = 0; i < indexOffsetLeg.abs(); i++) {
      if (indexOffsetLeg > 0) {
        ////当前item 左上移动
        if (newIndex + 1 * i == currentIndex) {
          _calOffsetLeftUp();
          break;
        }
      } else if (indexOffsetLeg < 0) {
        //当前item 右下移动
        if (newIndex - 1 * i == currentIndex) {
          _calOffsetRightBottom();
          break;
        }
      }
    }
  }

  //计算 item 向左上 要偏移的位置 然后开始动画
  _calOffsetLeftUp() async {
    RenderObject? renderObject = context.findRenderObject();

    var row = currentIndex % widget.crossAxisCount; //列数

    var itemWidth =
        (renderObject as RenderBox).size.width + widget.crossAxisSpacing;
    var itemHeight = 0.0;

    //判断列数是不是在最右边，最右边要向下做移动
    if (row == widget.crossAxisCount - 1) {
      itemWidth =
          -((renderObject as RenderBox).size.width + widget.crossAxisSpacing) *
              (widget.crossAxisCount - 1);
      itemHeight =
          (renderObject as RenderBox).size.height + widget.mainAxisSpacing;
    }
    _startOffset = _targetOffset;
    _targetOffset = Offset(itemWidth, itemHeight) + _targetOffset;
    startAnimate();

    await Future.delayed(const Duration(milliseconds: 50), () {
      currentIndex = currentIndex + 1;
    });
  }

  //计算 item 向右下上 要偏移的位置 然后开始动画
  _calOffsetRightBottom() async {
    if (!mounted) return;
    RenderObject? renderObject = context.findRenderObject();
    if (renderObject == null) {
      Offset.zero;
    }

    var row = currentIndex % widget.crossAxisCount; //列数

    var itemWidth =
        -((renderObject as RenderBox).size.width + widget.crossAxisSpacing);
    var itemHeight = 0.0;

    //判断列数是不是在最左边，最左边要向上做移动
    if (row == 0) {
      itemWidth =
          ((renderObject as RenderBox).size.width + widget.crossAxisSpacing) *
              (widget.crossAxisCount - 1);
      itemHeight =
          -((renderObject as RenderBox).size.height + widget.mainAxisSpacing);
    }
    _startOffset = _targetOffset;
    _targetOffset = Offset(itemWidth, itemHeight) + _targetOffset;
    startAnimate();

    await Future.delayed(const Duration(milliseconds: 50), () {
      currentIndex = currentIndex - 1;
    });
  }

  //开始动画
  startAnimate() {
    if (_offsetAnimation == null) {
      _offsetAnimation = AnimationController(vsync: this)
        ..duration = Duration(milliseconds: widget.animationDuration)
        ..addListener(rebuild)
        ..addStatusListener((status) {
          if (status == AnimationStatus.completed) {
            _offsetAnimation?.dispose();
            _offsetAnimation = null;
          }
        })
        ..forward(from: 0.0);
    } else {
      _startOffset = offset;
      _offsetAnimation?.forward(from: 0.0);
    }
  }

  //item 偏移量
  Offset get offset {
    if (_offsetAnimation != null) {
      return Offset.lerp(
        _startOffset,
        _targetOffset,
        Curves.easeInOut.transform(_offsetAnimation!.value),
      )!;
    }
    return _targetOffset;
  }

  //删除item时其他 item的位移
  deleteItemOffset() {
    _calOffsetRightBottom();
  }

  isDelete(bool isAnimate) {
    this.isAnimate = isAnimate;
  }

  @override
  Widget build(BuildContext context) {
    return GestureDetector(
      onTapDown: widget.isOpenDragAble ? onTapDownEvent : null,
      onPanUpdate: widget.isOpenDragAble ? onPanUpdateEvent : null,
      onPanEnd: widget.isOpenDragAble ? onPanEndEvent : null,
      onTapUp: !isAnimate ? onTapUpEvent : null,
      child: _buildChild(),
    );
  }

  Widget _buildChild() {
    if (_dragging) {
      return SizedBox();
    }

    return Transform(
      transform: Matrix4.translationValues(offset.dx, offset.dy, 0),
      child: Stack(
        fit: StackFit.passthrough,
        clipBehavior: Clip.none,
        children: [
          Container(
            child: widget.itemChild,
          ),
          widget.isActive
              ? Positioned(
                  top: -2,
                  right: -2,
                  child: widget.tRWidget,
                )
              : SizedBox(),
        ],
      ),
    );
  }

  //按下事件
  onTapDownEvent(TapDownDetails e) async {
    if (isAnimate) return;
    isLongPress = true;

    await Future.delayed(Duration(milliseconds: widget.longPressDuration));
    if (isLongPress == true) {
      //震动
      Vibration.vibrate(duration: 15, amplitude: 100);
      if (widget.isActive) {
        _listState.dragStart(
          widget.index,
          context,
          e,
          widget.itemChild,
          widget.tRWidget,
          widget.crossAxisCount,
          widget.crossAxisSpacing,
          widget.mainAxisSpacing,
        );

        _dragging = true;

        rebuild();
      } else {
        widget.gestureCallback.onActive(true);
      }
    }
  }

  //拖动事件
  onPanUpdateEvent(e) {
    if (_dragging == true) {
      _listState.dragUpdate(e);
    }
  }

  //拖动结束事件
  onPanEndEvent(e) {
    _listState.dragCancel(currentIndex, newIndex);
  }

  //点击事件
  onTapUpEvent(e) async {
    isLongPress = false;

    if (isAnimate) return;

    if (_dragging) {
      _dragging = false;

      _listState.dragCancel(currentIndex, newIndex);

      rebuild();
    } else {
      if (widget.isActive) {
        //走删除

        _dragging = true;

        _listState.isDeleteAnimate(true);

        _listState.deleteItem(currentIndex, widget.fixedNum);

        rebuild();

        await Future.delayed(
            Duration(milliseconds: widget.animationDuration + 10), () {
          // 等待动画结束  删除回调 偷个懒 直接一个计时加10

          _listState.isDeleteAnimate(false);

          widget.gestureCallback.onTapDelete(currentIndex);
        });
      } else {
        //走点击
        widget.gestureCallback.onTapItem(currentIndex);
      }
    }
  }

  //拖拽结束回调
  onPanEndBack() {
    if (!_dragging) return;

    isLongPress = false;
    _dragging = false;
    setState(() {});

    if (newIndex == -1) return;
    widget.gestureCallback.onPanEnd(currentIndex, newIndex);

    newIndex = -1;
  }

  void rebuild() {
    if (mounted) {
      setState(() {});
    }
  }
}


lib/drag_gridview_mixin.dart
import 'dart:async';
import 'dart:math';

import 'package:flutter/cupertino.dart';
import 'package:animated_drag_gridview/drag_info.dart';
import 'package:animated_drag_gridview/drag_item.dart';
import 'package:flutter/foundation.dart';

mixin DragGridViewMixin<T extends StatefulWidget> on State<T> {
  static DragGridViewMixin of(BuildContext context) {
    return context.findAncestorStateOfType<DragGridViewMixin>()!;
  }

  final Map<int, DragItemState> __dragItems = <int, DragItemState>{};
  final Map<int, DragInfoState> __dragInfo = <int, DragInfoState>{};

  void dragItem(DragItemState item) {
    __dragItems[item.index] = item;
  }

  void dragInfo(DragInfoState item) {
    __dragInfo[0] = item;
  }

  Offset dragStartPos = Offset.zero; //开始拖拽时的 手指位置
  Offset dragInfoOffset = Offset.zero; //开始拖拽时 拖拽小组件的偏移量
  int dragIndex = -1;
  double itemWidth = 0;
  double itemHeight = 0;

  int crossAxisCount = 3;
  double mainAxisSpacing = 0;
  double crossAxisSpacing = 0;

  //拖拽开始
  void dragStart(
    int dragIndex,
    BuildContext context,
    TapDownDetails e,
    Widget itemChild,
    Widget tRWidget,
    int crossAxisCount,
    double mainAxisSpacing,
    double crossAxisSpacing,
  ) {
    dragStartPos = Offset(e.localPosition.dx, e.localPosition.dy);
    this.dragIndex = dragIndex;
    this.crossAxisCount = crossAxisCount;
    this.mainAxisSpacing = mainAxisSpacing;
    this.crossAxisSpacing = crossAxisSpacing;

    RenderObject? renderObject = context.findRenderObject();
    if (renderObject == null) {
      Offset.zero;
    }
    //  var cent = (renderObject as RenderBox).size.center(Offset.zero);
    itemWidth = (renderObject as RenderBox).size.width;
    itemHeight = (renderObject as RenderBox).size.height;

    dragInfoOffset = calOffset(dragIndex);

    //通知拖拽小部件
    __dragInfo[0]?.dragInfoStart(
      dragInfoOffset,
      itemChild,
      tRWidget,
      itemWidth,
      itemHeight,
    );
  }

  ///计算拖拽小组件的开始 偏移量
  Offset calOffset(index) {
    var row = index % crossAxisCount; //列数
    var col = (index / crossAxisCount).floor(); //行数
    var offsetX = row * itemWidth + mainAxisSpacing * row; //列坐标
    var offsetY = col * itemHeight + crossAxisSpacing * col; //行坐标
    return Offset(offsetX, offsetY);
  }

  //拖拽中
  void dragUpdate(DragUpdateDetails e) {
    Offset updateOffset = Offset(
        dragInfoOffset.dx + e.localPosition.dx - dragStartPos.dx,
        dragInfoOffset.dy + e.localPosition.dy - dragStartPos.dy);

    //通知小组件开始更新位置
    __dragInfo[0]?.dragInfoUpdate(updateOffset);

    _calcDragItemIsAnimate(updateOffset);
  }

  //拖拽取消
  void dragCancel(currentIndex, toIndex) {
    if (toIndex != -1) {
      dragInfoOffset = calOffset(toIndex);
    }

    __dragInfo[0]?.dragInfoCancel(currentIndex, toIndex, dragInfoOffset);

    dragInfoOffset = Offset(0.0, 0.0);
    dragStartPos = Offset(0.0, 0.0);
  }

  //拖拽取消回调
  void dragCancalBack() {
    for (var item in __dragItems.values) {
      item.onPanEndBack();
    }
  }

  //点击item 删除item
  void deleteItem(currentIndex, fixedNum) {
    for (int i = currentIndex + 1; i < __dragItems.length + fixedNum; i++) {
      __dragItems[i]?.deleteItemOffset();
    }
    __dragItems.clear();
  }

  void isDeleteAnimate(bool isAnimate) {
    for (var item in __dragItems.values) {
      item.isDelete(isAnimate);
    }
  }

  ///计算 DragItem 是否需要排序动画
  void _calcDragItemIsAnimate(Offset updateOffset) {
    for (var item in __dragItems.values) {
      var dragNewItemIndex = item.index;

      Offset otherItem = calOffset(dragNewItemIndex);

      var ah = sqrt(pow(updateOffset.dx - otherItem.dx, 2) +
          pow(updateOffset.dy - otherItem.dy, 2));

      if (ah < itemWidth / 2 && ah < itemHeight / 2) {
        if (dragNewItemIndex != dragIndex) {
          for (var itemf in __dragItems.values) {
            //dragNewItemIndex 移动到新的 item 的 index
            itemf.updateForGap(dragIndex, dragNewItemIndex);
          }
          dragIndex = dragNewItemIndex;
          break;
        }
      }
    }
  }
}


lib/animated_drag_gridview.dart
import 'dart:ffi';

import 'package:animated_drag_gridview/drag_gridview_mixin.dart';
import 'package:animated_drag_gridview/drag_info.dart';
import 'package:animated_drag_gridview/drag_item.dart';
import 'package:animated_drag_gridview/single_touch_recognizer.dart';
import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';
import 'package:flutter/src/foundation/key.dart';
import 'package:flutter/src/widgets/framework.dart';

typedef ReorderCallback = void Function(int oldIndex, int newIndex);

typedef DeleteItemCallback = void Function(int currentIndex);

typedef ActiveCallback = void Function(bool isActive);

typedef OnTapItemCallback = void Function(int index);

class AnimateDragGridView extends StatefulWidget {
  int crossAxisCount;
  double mainAxisSpacing;
  double crossAxisSpacing;
  double childAspectRatio;
  ScrollPhysics? physics;

  ///动画时长 milliseconds
  final int animationDuration;

  ///长安时长 milliseconds
  final int longPressDuration;

  //item
  List<Widget> children = const <Widget>[];

  //GridView 的padding
  EdgeInsetsGeometry padding;

  ///是否可以拖拽
  final bool isOpenDragAble;

  ///是否激活
  final bool isActive;

  ///固定前几个标签
  final int fixedNum;

  ///激活item 时右上角 显示的 小部件
  Widget tRWidget;

  //编辑回调
  ActiveCallback? onActive;

  ///重新排序 回调
  final ReorderCallback onReorder;

  ///删除item 回调
  final DeleteItemCallback? onDeleteItem;

  /// 未激活时 点击item 的回调 可做跳转处理
  final OnTapItemCallback? onTapItem;

  AnimateDragGridView({
    this.physics,
    required this.children,
    this.onActive,
    required this.onReorder,
    this.onDeleteItem,
    this.onTapItem,
    this.mainAxisSpacing = 0,
    this.crossAxisSpacing = 0,
    this.childAspectRatio = 1.0,
    this.crossAxisCount = 3,
    this.padding = const EdgeInsets.all(0),
    this.isOpenDragAble = false,
    this.isActive = false,
    this.fixedNum = 0,
    this.tRWidget = const SizedBox(),
    this.animationDuration = 350,
    this.longPressDuration = 600,
    Key? key,
  }) : super(key: key);

  @override
  State<AnimateDragGridView> createState() => _AnimateDragGridViewState();
}

class _AnimateDragGridViewState extends State<AnimateDragGridView>
    with DragGridViewMixin
    implements GestureCallback {
  @override
  void onPanEnd(int oldIndex, int newIndex) {
    // TODO: implement onPanEnd
    widget.onReorder(oldIndex, newIndex);
  }

  @override
  void onTapDelete(int currentIndex) {
    // TODO: implement onTapDelete
    if (widget.onDeleteItem == null) return;
    widget.onDeleteItem!(currentIndex);
  }

  @override
  void onActive(bool isActive) {
    // TODO: implement onTapDown
    if (widget.onActive == null) return;
    widget.onActive!(isActive);
  }

  @override
  void onTapItem(int index) {
    // TODO: implement onTapItem
    if (widget.onTapItem == null) return;
    widget.onTapItem!(index);
  }

  @override
  Widget build(BuildContext context) {
    return SingleTouchRecognizerWidget(
      child: Padding(
        padding: widget.padding,
        child: Stack(
          clipBehavior: Clip.none,
          children: [
            GridView.builder(
              clipBehavior: Clip.none,
              shrinkWrap: true,
              physics: widget.physics,
              gridDelegate: SliverGridDelegateWithFixedCrossAxisCount(
                crossAxisCount: widget.crossAxisCount,
                crossAxisSpacing: widget.crossAxisSpacing,
                mainAxisSpacing: widget.mainAxisSpacing,
                childAspectRatio: widget.childAspectRatio,
              ),
              itemCount: widget.children.length,
              itemBuilder: (BuildContext context, int index) {
                return item(index);
              },
            ),
            DragInfo(animationDuration: widget.animationDuration)
          ],
        ),
      ),
    );
  }

  Widget item(int index) {
    if (index >= widget.fixedNum) {
      return DragItem(
        key: GlobalKey(),
        index: index,
        itemChild: widget.children[index],
        gestureCallback: this,
        crossAxisCount: widget.crossAxisCount,
        crossAxisSpacing: widget.crossAxisSpacing,
        mainAxisSpacing: widget.mainAxisSpacing,
        isOpenDragAble: widget.isOpenDragAble,
        isActive: widget.isActive,
        fixedNum: widget.fixedNum,
        tRWidget: widget.tRWidget,
        animationDuration: widget.animationDuration,
        longPressDuration: widget.longPressDuration,
      );
    }

    return GestureDetector(
      onTap: widget.isActive
          ? null
          : () {
              if (widget.onTapItem == null) return;
              widget.onTapItem!(index);
            },
      child: Opacity(
        opacity: widget.isActive ? 0.5 : 1,
        child: widget.children[index],
      ),
    );
  }
}

//手势回调
abstract class GestureCallback {
  void onTapDelete(int index);
  void onActive(bool isActive);
  void onPanEnd(int dragIndex, int toIndex);
  void onTapItem(int index);
}


example/lib/drag_gridview.dart
import 'package:flutter/material.dart';
import 'package:animated_drag_gridview/animated_drag_gridview.dart';

class DragGridView extends StatefulWidget {
  DragGridView({Key? key}) : super(key: key);

  @override
  State<DragGridView> createState() => _DragGridViewState();
}

class _DragGridViewState extends State<DragGridView> {
  bool isActive = false;
  @override
  void initState() {
    super.initState();
  }

  List items = [
    "张飞",
    '关羽',
    '李白',
    '刘备',
    '周瑜',
    '扁鹊',
    '上官',
    '苏烈',
    '元哥',
    '曹操',
    '廉颇',
    '夏侯',
    '梦琪',
  ];
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.white,
      appBar: AppBar(
        centerTitle: true,
        backgroundColor: Colors.white,
        elevation: 0.5,
        title: Text("频道设置", style: TextStyle(color: Colors.black)),
      ),
      body: Column(
        children: [
          topEdit(),
          AnimateDragGridView(
            crossAxisCount: 4,
            childAspectRatio: 2.2,
            crossAxisSpacing: 8.0,
            mainAxisSpacing: 8.0,
            isOpenDragAble: true,
            isActive: isActive,
            fixedNum: 3,
            animationDuration: 350, //milliseconds
            longPressDuration: 600, //milliseconds
            padding: const EdgeInsets.symmetric(horizontal: 15),
            physics: const NeverScrollableScrollPhysics(),
            children: items.map((e) => buildItem(e)).toList(),
            tRWidget: tRWidget(),
            onReorder: (oldIndex, newIndex) {
              var removeItem = items.removeAt(oldIndex);
              items.insert(newIndex, removeItem);
              setState(() {});
            },
            onDeleteItem: (index) {
              items.removeAt(index);
              setState(() {});
            },
            onActive: (isActive) {
              this.isActive = isActive;
              setState(() {});
            },
            onTapItem: (index) {
              print("点击的item");
              print(index);
            },
          ),
        ],
      ),
    );
  }

  Widget tRWidget() {
    return Container(
      height: 15,
      width: 15,
      alignment: Alignment.center,
      decoration: BoxDecoration(
        color: Color.fromARGB(255, 208, 208, 220),
        borderRadius: BorderRadius.circular(50),
      ),
      child: Icon(
        Icons.remove,
        color: Colors.white,
        size: 13,
      ),
    );
  }

  Widget buildItem(e) {
    return Container(
      decoration: BoxDecoration(
        color: Color.fromARGB(255, 231, 236, 245),
        borderRadius: BorderRadius.circular(2),
      ),
      alignment: Alignment.center,
      child: Text(
        e.toString(),
        style: TextStyle(
          color: Color.fromARGB(255, 86, 86, 90),
        ),
      ),
    );
  }

  Widget topEdit() {
    return Container(
      padding: EdgeInsets.all(15),
      child: Row(
        crossAxisAlignment: CrossAxisAlignment.end,
        children: [
          Text(
            "我的频道",
            style: TextStyle(
                fontSize: 17, fontWeight: FontWeight.w600, color: Colors.black),
          ),
          SizedBox(width: 5),
          Text(
            isActive ? "长按可拖拽排序" : "点击进入频道",
            style: TextStyle(fontSize: 15, color: Colors.black26),
          ),
          Spacer(),
          InkWell(
            child: Text(
              isActive ? "完成" : "编辑",
              style: TextStyle(fontSize: 15, color: Colors.blue),
            ),
            onTap: () {
              isActive = !isActive;
              setState(() {});
            },
          ),
        ],
      ),
    );
  }
}
